kad-dht/README.md

```diff
@@ -0,0 +1,439 @@
+	// newest).
+	//
+	// Decisions made by select should be stable.
+	Select(key string, values [][]byte) (int, error)
+}
+```
+
+`Validate()` should be a pure function that reports the validity of a record. It
+may validate a cryptographic signature, or else. It is called on two occasions:
+
+1. To validate values retrieved in a `GET_VALUE` query.
+2. To validate values received in a `PUT_VALUE` query before storing them in the
+   local data store.
+
+Similarly, `Select()` is a pure function that returns the best record out of 2
+or more candidates. It may use a sequence number, a timestamp, or other
+heuristic of the value to make the decision.
+
+### Content provider advertisement and discovery
+
+Nodes must keep track of which nodes advertise that they provide a given key
+(CID). These provider advertisements should expire, by default, after 24 hours.
+These records are managed through the `ADD_PROVIDER` and `GET_PROVIDERS`
+messages.
+
+#### Content provider advertisement
+
+When the local node wants to indicate that it provides the value for a given
+key, the DHT finds the closest peers to the key using the `FIND_NODE` RPC (see
+[peer routing section](#peer-routing)), and then sends an `ADD_PROVIDER` RPC with
+its own `PeerInfo` to each of these peers.
+
+Each peer that receives the `ADD_PROVIDER` RPC should validate that the received
+`PeerInfo` matches the sender's `peerID`, and if it does, that peer should store
+the `PeerInfo` in its datastore. Implementations may choose to not store the
+addresses of the providing peer e.g. to reduce the amount of required storage or
+to prevent storing potentially outdated address information.
+
+#### Content provider discovery
+
+_Getting_ the providers for a given key is done in the same way as _getting_ a
+value for a given key (see [getting values section](#getting-values)) except
+that instead of using the `GET_VALUE` RPC message the `GET_PROVIDERS` RPC
+message is used.
+
+When a node receives a `GET_PROVIDERS` RPC, it must look up the requested
+key in its datastore, and respond with any corresponding records in its
+datastore, plus a list of closer peers in its routing table.
+
+### Bootstrap process
+
+The bootstrap process is responsible for keeping the routing table filled and
+healthy throughout time. It runs once on startup, then periodically with a
+configurable frequency (default: 5 minutes).
+
+On every run, we generate a random peer ID and we look it up via the process
+defined in [peer routing](#peer-routing). Peers encountered throughout the
+search are inserted in the routing table, as per usual business.
+
+This is repeated as many times per run as configuration parameter `QueryCount`
+(default: 1). In addition, to improve awareness of nodes close to oneself,
+implementations should include a lookup for their own peer ID.
+
+Every repetition is subject to a `QueryTimeout` (default: 10 seconds), which
+upon firing, aborts the run.
```

Comment:
Should include our usual, here's one simple way of doing this. Especially since neither go nor rust libp2p's implementations actually do this.

Reply:
:+1: added via https://github.com/libp2p/specs/pull/108/commits/3e6f8f5e9a98b7078fb9ae7020179aa7d2c86947.